crates/subspace-core-primitives/src/lib.rs

```diff
@@ -279,17 +281,49 @@ impl PotProof {
     }
 
     /// Returns the last check point.
-    pub fn output(&self) -> Option<PotCheckpoint> {
-        self.checkpoints.last().cloned()
+    pub fn output(&self) -> PotCheckpoint {
+        self.checkpoints.last()
     }
 
     /// Derives the global randomness from the output.
-    pub fn derive_global_randomness(&self) -> Option<Blake2b256Hash> {
-        self.output()
-            .map(|checkpoint| blake2b_256_hash(&PotBytes::from(checkpoint)))
+    pub fn derive_global_randomness(&self) -> Blake2b256Hash {
+        blake2b_256_hash(&PotBytes::from(self.output()))
+    }
+
+    /// Derives the next seed based on the injected randomness.
+    pub fn next_seed(&self, injected_hash: Option<BlockHash>) -> PotSeed {
+        match injected_hash {
+            Some(injected_hash) => {
+                // Next seed = Hash(last checkpoint + injected hash).
+                let mut bytes = self.output().0.to_vec();
+                bytes.append(&mut injected_hash.to_vec());
+                PotSeed::from(pot_bytes_from_hash_256(blake2b_256_hash(&bytes)))
+            }
+            None => {
+                // No injected randomness, next seed = last checkpoint.
+                PotSeed::from(self.output().0)
+            }
+        }
+    }
+
+    /// Derives the next key from the hash of the current seed.
+    pub fn next_key(&self) -> PotKey {
+        PotKey::from(pot_bytes_from_hash_256(blake2b_256_hash(&self.seed.0)))
     }
 }
 
+/// Helper to convert the 32 byte hash to 16 bytes.
+fn pot_bytes_from_hash_256(hash: Blake2b256Hash) -> PotBytes {
+    let mut h: [u8; 16] = Default::default();
+    h.copy_from_slice(&hash[0..16]);
+
+    let mut l: [u8; 16] = Default::default();
+    l.copy_from_slice(&hash[16..32]);
+
+    let r = u128::from_be_bytes(h) ^ u128::from_be_bytes(l);
```

Comment:
You can really just truncate here and take first 16 bytes. The XOR is not needed.